File: Poker.py
#
'''This Python program solves Project Euler problem #54.
It reads in a file with two poker hands per line and determines
how many are won by hand one. There are no ties.

Much of the code is not as efficient as could be. It was written
with understanding and maintainability in mind. Parts of the code
are described in the adjacent file outline.vim
'''
__author__ = 'rzucker'

from collections import Counter
from operator import itemgetter

CARD_VALS = "_123456789TJQKA" # Used for converting the card values to 2..14
SUIT_VALS = "CDHS"            # Used for converting suits to 0..3

# Used to be a constant for hand types for comparison.
# This dictionary isn't truly necessary, but it makes for easier
# to understand code.
HAND_TYPE_VALS = {'Straight Flush': 9,
                  '4 Kind':         8,
                  'Full House':     7,
                  'Flush':          6,
                  'Straight':       5,
                  '3 Kind':         4,
                  'Two Pair':       3,
                  'One Pair':       2,
                  'High Card':      1}

# This dictionary is just used to make it easier to code debug messages when needed
DEBUG_HAND_TYPE_STRING = {9:"Straight Flush",
                          8:"4 of a Kind",
                          7:"Full House",
                          6:"Flush",
                          5:"Straight",
                          4:"3 of a Kind",
                          3:"Two Pair",
                          2:"One Pair",
                          1:"High Card"}

def straight(hand):
    '''Returns true if hand is a straight

    hand is a two dimension array, 5x2, with a single hand
    hand is assumed to already sorted by value from high to low
    Therefore, can determine if it is a straight by making sure next card is
    one less than previous
    '''

    # last_card is set to highest
    last_card = hand[0][0]
    for i in range(1, 5):
        # for next four cards make sure it is one less than the previous card
        if (last_card-1) == hand[i][0]:
            # set last card to be the latest card checked
            last_card = hand[i][0]
        else:
            return False
    return True


def flush(hand):
    '''Returns true if hand is a flush; i.e., suit is same for all five cards
    hand is a two dimension array, 5x2, with a single hand
    '''
    # set suit to that of the first card
    suit = hand[0][1]
    for i in range(1, 5):
        # check suit of each remaining card to that of the first card
        if suit != hand[i][1]:
            return False
    return True


def straight_flush(hand):
    '''Call the straight and flush functions.
    Could save computation by using a combined function
    to figure out all three, but this is clearer
    '''
    return straight(hand) and flush(hand)


def card_val_convert(card):
    '''Convert the card value into an integer, two to fourteen'''
    card_val = CARD_VALS.find(card)
    return card_val


def card_suit_convert(card):
    '''Convert the suit into an integer, zero to three'''
    card_suit_num = SUIT_VALS.find(card)
    return card_suit_num


def hand_type(hand, card_counts):
    '''Determine the type of hand.
    Leverage straight() and flush() functions
    Also leverage count of most frequently occurring pip value in card_counts
    '''
    unique_vals = len(card_counts)
    # If there are five unique card values, then it is one of straight, flush, straight flush, or just high card
    if unique_vals == 5:
        if straight_flush(hand):
            return HAND_TYPE_VALS['Straight Flush']
        elif flush(hand):
            return HAND_TYPE_VALS['Flush']
        elif straight(hand):
            return HAND_TYPE_VALS['Straight']
        else:
            return HAND_TYPE_VALS['High Card']
    # If there are four unique card values, can only be one pair
    elif unique_vals == 4:
        return HAND_TYPE_VALS['One Pair']
    # If there are three unique card values, then either three of a kind or two pair
    elif unique_vals == 3:
        # If most frequent unique value occurs three times, then three of a kind
        if card_counts[0][1] == 3:
            return HAND_TYPE_VALS['3 Kind']
        else:
            return HAND_TYPE_VALS['Two Pair']
    # If there are two unique card values, it is four of a kind or a full house (three and two of a kind)
    elif unique_vals == 2:
        # If most frequent unique value occurs four times, then four of a kind
        if card_counts[0][1] == 4:
            return HAND_TYPE_VALS['4 Kind']
        else:
            return HAND_TYPE_VALS['Full House']
    # Without wild cards you cannot have five of a kind
    else:
        print("Bad value for number of unique values", unique_vals)


def two_pair_tie_breaker(counts_card1, counts_card2):
    '''Determining the winner when both hands have two pair is complicated enough to needs its own function'''
    # Check to see who has the highest two pair
    if counts_card1[0][0] > counts_card2[0][0]:
        return 1
    elif counts_card2[0][0] > counts_card1[0][0]:
        return 2
    # Top pair is equal. Check second pair
    elif counts_card1[1][0] > counts_card2[1][0]:
        return 1
    elif counts_card2[1][0] > counts_card1[1][0]:
        return 2
    # Both pair are equal. Check fifth card (third in count array)
    elif counts_card1[2][0] > counts_card2[2][0]:
        return 1
    elif counts_card2[2][0] > counts_card1[2][0]:
        return 2
    else:
        print("All cards match in two pairs. Illegal input")
        return -1

def one_pair_tie_breaker(hand1, hand2, counts_card1, counts_card2):
    '''Determining the winner when both hands have a pair is complicated enough to needs its own function'''
    if counts_card1[0][0] > counts_card2[0][0]:
        return 1
    elif counts_card2[0][0] > counts_card1[0][0]:
        return 2
    # Pair has same value, check highest of hand one by one until mismatch
    else:
        for i in range(0, 5):
            if hand1[i][0] > hand2[i][0]:
                return 1
            elif hand2[i][0] > hand1[i][0]:
                return 2
        # All cards equal
        print("All cards equal. Illegal input")
        return -1


def high_card_tie_breaker(hand1, hand2):
    '''Determining the winner when just high card has enough steps to needs its own function'''
    for i in range(0, 5):
        if hand1[i][0] > hand2[i][0]:
            return 1
        elif hand2[i][0] > hand1[i][0]:
            return 2
    # All cards equal
    print("All cards equal. Illegal input")
    return -1


def tie_breaker(type_of_hand, hand1, hand2, counts_card1, counts_card2):
    '''Hand type is the same for both hands. Determine who wins via tie breaker routines
    Params: hand type, hand1 and hand2 (2D array), and sorted card count array for hand1 and hand2
    '''
    # If the hand is a flush, straight, or straight flush
    if type_of_hand == 9 or type_of_hand == 6 or type_of_hand == 5:
        # Can determine by comparing the highest card in each hand
        if hand1[0][0] > hand2[0][0]:
            print("Hand1 wins")
            return 1
        elif hand2[0][0] > hand1[0][0]:
            return 2
        else:
            print("Tie in hands - bad data")
            return -1
    # Hand is four of a kind
    elif type_of_hand == 8:
        # Card values in a four of a kind are unique. Just compare them
        if counts_card1[0][0] > counts_card2[0][0]:
            return 1
        else:
            return 2
    # Hand is full house or three of kind
    elif type_of_hand == 4 or type_of_hand == 7:
        # Can determine winner by comparing value of three of a kind. Must be different.
        if counts_card1[0][0] > counts_card2[0][0]:
            return 1
        else:
            return 2
    elif type_of_hand == 3:
        return two_pair_tie_breaker(counts_card1, counts_card2)
    elif type_of_hand == 2:
        return one_pair_tie_breaker(hand1, hand2, counts_card1, counts_card2)
    elif type_of_hand == 1:
        return high_card_tie_breaker(hand1, hand2)
    else:
        print("Illegal hand type (over nine or less than one")
        return -1

def main():
    '''Determine winning poker hand from two hands read from file.'''
    hand1_wins = 0
    for line in open('poker.txt'):
	# Use rstrip() to get rid of \n at the end
        hands = line.rstrip().split(" ")
	# Create a two dimensional array for hand1 and hand2
	# One for suit, one for value for each of five cards in a hand
        hand1 = [[0, 0]] * 5
        hand2 = [[0, 0]] * 5
        # Need a one dimensional array for the pip values of the cards in each hand
        hand1_vals = [0] * 5
        hand2_vals = [0] * 5
	# Convert each hand into numeric form
        for i in range(5):
            # Convert the text form of the pip values to numbers, two to 14
            hand1[i][0] = card_val_convert(hands[i][0])
            hand2[i][0] = card_val_convert(hands[i+5][0])
            # Convert values of cards into a one dimensional list
            hand1_vals[i] = hand1[i][0]
            hand2_vals[i] = hand2[i][0]
            # Convert card suit to numeric values, zero to three
            hand1[i][1] = card_suit_convert(hands[i][1])
            hand2[i][1] = card_suit_convert(hands[i+5][1])
        hand1.sort(reverse=True)
        hand2.sort(reverse=True)
	# Create a counter (hash table) that counts how many of each card value in each hand
        cards1 = Counter()
        cards2 = Counter()
        for each_card in hand1_vals:
            # Increment count of hash table value for card values in hand1
            cards1[each_card] += 1
        for each_card in hand2_vals:
            # Increment count of hash table value for card values in hand2
            cards2[each_card] += 1
        # Convert the counter into a dictionary, and sort by most common appearing,
	# card value and return as a list
        counts_card1 = sorted(cards1.items(), key=itemgetter(1), reverse=True)
        counts_card2 = sorted(cards2.items(), key=itemgetter(1), reverse=True)
        # print("Hand 1 is", hand1, ", Hand 2 is", hand2)
        print("Hand 1 type is", DEBUG_HAND_TYPE_STRING[hand_type(hand1, counts_card1)], "   "
               "Hand 2 type is", DEBUG_HAND_TYPE_STRING[hand_type(hand2, counts_card2)])
        if hand_type(hand1, counts_card1) > hand_type(hand2, counts_card2):
            print("Hand 1 wins")
            hand1_wins += 1
        elif hand_type(hand1, counts_card1) == hand_type(hand2, counts_card2):
            if tie_breaker(hand_type(hand1, counts_card1), hand1, hand2, counts_card1, counts_card2) == 1:
                print("Hand 1 wins")
                hand1_wins += 1
    print("Hand 1 wins", hand1_wins, "times.")



# Check for interactive session
if __name__ == '__main__':
    # execute main program
    main()

File: C/poker.c
/* This C program solves Project Euler problem #54.
        
        Program reads in a file with two poker hands per line and determines
        how many are won by hand one. There are no ties.
        
*/

#include <stdio.h>
#include <stdbool.h>
#include <string.h>
#include <stdlib.h>

/* Don't really need an enum. Could just use the char (C|D|H|S) that is read in
   from the file. Just an excuse to demonstrate that I know how to use enums. */
enum card_suits { clubs, diamonds, hearts, spades };

typedef struct {
           int  pip;
           enum card_suits suit;
} t_card;

typedef struct {
           int pip;
           int count;
} t_card_count;
        
char* CARD_VALS = "0123456789TJQKA";

/* Define the hand types, from highest to lowest */
#define STRAIGHT_FLUSH  9
#define FOUR_KIND       8
#define FULL_HOUSE      7
#define FLUSH           6
#define STRAIGHT        5
#define THREE_KIND      4
#define TWO_PAIR        3
#define ONE_PAIR        2
#define HIGH_CARD       1

int hand_pip_compare (first_elem, second_elem)
    const void *first_elem, *second_elem;
    {
        /* Do a descending sort */
        int first, second;

        first  = (*(t_card *) first_elem).pip;
        second = (*(t_card *) second_elem).pip;
        if (first < second)
            return 1;
        else if (first > second)
            return -1;
        else
            return 0;
    }

int card_count_pip_compare (first_elem, second_elem)
    const void *first_elem, *second_elem;
    {
        int first, second, pip1, pip2;

        first  = (*(t_card_count *) first_elem).count;
        second = (*(t_card_count *) second_elem).count;
        pip1   = (*(t_card_count *) first_elem).pip;
        pip2   = (*(t_card_count *) second_elem).pip;

        if (first < second)
            return 1;
        else if (first > second)
            return -1;
        /* same occurence count, but higher pip value */
        else if (pip1 < pip2)
            return 1;
        else if (pip1 > pip2)
            return -1;
        else if ((first == 0) && (pip1 == 0))
            /* Same values, but initialized zero values */
            return 0;
        else {
            /* same pip value. Mistake, should have been merged in */
            printf ("Error card_count_pip_compare: same pip value twice: %d\n", pip1);
            return 0;
        }
    }

int count_pips (hand, pip_list)
    t_card *hand;
    t_card_count *pip_list;
    /* Fill in the pip_list array with the count of each pip in the hand.
       Return the number of distinct pip values. */
    {
        int i, j;

        j = 0; /* Count of where we are in pip_list */
        pip_list [0].pip   = hand [0].pip;
        pip_list [0].count = 1;
        /* Clear pip list, start at 1 since already assigned zero */
        for (i = 1; i < 5; i++) {
            pip_list [i].pip   = 0;
            pip_list [i].count = 0;
        }
        for (i = 1; i < 5; i++) {
            /* Could use a nested loop, but with the fixed number, just went for if-else */
            if (hand [i].pip == pip_list [0].pip) {
                pip_list [0].count++;
            } else if (hand [i].pip == pip_list [1].pip) {
                pip_list [1].count++;
            } else if (hand [i].pip == pip_list [2].pip) {
                pip_list [2].count++;
            } else if (hand [i].pip == pip_list [3].pip) {
                pip_list [3].count++;
            } else if (hand [i].pip == pip_list [4].pip) {
                pip_list [4].count++;
            } else { /* New pip value */
                j++;
                pip_list [j].pip   = hand [i].pip;
                pip_list [j].count = 1;
            }
        }
        /* Now sort based upon count */
        qsort (pip_list, 5, sizeof(t_card_count), card_count_pip_compare);
        return j + 1; /* Started at zero. Need to increment by one. */
    }

/* Returns true if hand is a straight. */
_Bool straight (hand)
    t_card *hand;
    {
        /* 
        :param hand: an array with a single hand of five elements
        hand is assumed to already sorted by value from high to low
        Therefore, can determine if it is a straight by making sure
        the next card is one less than previous
        */
        
        int i, last_card;
        /* last_card is set to highest */
        last_card = hand[0].pip;
        /* for next four cards make sure it is one less than previous card */
        for (i = 1; i < 5; i++) {
            if ((last_card-1) == hand[i].pip)
                /* set last card to be the latest card checked */
                last_card = hand[i].pip;
            else
                return false;
        }
        return true;
    }
        
/* Return true if hand is a flush, i.e., suit is same for all five cards */
_Bool flush(hand)
    t_card *hand;
    {
        /* :param hand: an array with a single hand of five elements */
        
        int i;
        enum card_suits suit;
        /* set suit to that of the first card */
        suit = hand[0].suit;
        for (i = 0; i < 5; i++) {
            /* check suit of each remaining card to that of the first card */
            if (suit != hand[i].suit) {
                return false;
            }
        }
        return true;
    }
        
        
_Bool straight_flush(hand)
    t_card *hand;
    {
        /* Call the straight() and flush() functions.
           :param hand: a 2D array, 5x2, with a single hand
           Could save computation by using a combined function to figure out
           all three, but this is clearer
        */
        
        return straight(hand) && flush(hand);
    }

int card_val_convert(pip_value)
    char *pip_value; /* a single string value, from 2 to TJQKA */
        /* Convert the card value into an integer, two to fourteen. */
    {
        char *finder;

        finder = strchr(CARD_VALS, *pip_value);
        return (int) (finder - CARD_VALS);
    }

enum card_suits card_suit_convert(card_suit)
    char card_suit; /* :param card: a single character value, one of C, D, H, or S */
    {
        
        switch (card_suit) {
            case 'C':
                return clubs; /* no break needed */
            case 'D':
                return diamonds;
            case 'H':
                return hearts;
            case 'S':
                return spades;
            default:
                printf ("Error card_suit_convert: Bad input for suit %c", card_suit);
                return (enum card_suits) -1;
        }
    }

int num_vals (card_counts, max_cnt)
    int card_counts [14];
    int *max_cnt;
    {
        /* Count how many unique values are in a hand using the card_counts array
           Also how many times does the most frequent pip value occur */
        int i, count;

        count    = 0;
        *max_cnt = 0;
        for (i = 1; i < 14; i++) {
            if (card_counts [i] != 0) {
                count++;
                if (card_counts [i] > *max_cnt) {
                    *max_cnt = card_counts [i];
                }
            }
        }
        if (count > 5)
            printf ("Error: %d unique values in a hand", count);
        return (count);
    }

int hand_type (hand, unique_vals, how_often)
    t_card hand [5];
    int unique_vals;
    int how_often;
    {
        /* Determines the type of hand.
        :param hand: an array of five elments of type card
        :param unique_vals: the number of unique pip values in the hand
        :param how_often: how often the most frequent pip value occurs
        :return: encodedhand type values (defined using #define)
        Leverages straight() and flush() functions
        */
        /* If there are five unique card values, then it is one of straight, flush,
           straight flush, or just high card */
        if (unique_vals == 5) {
            if (straight_flush (hand))
                return (STRAIGHT_FLUSH);
            else if (flush (hand))
                return (FLUSH);
            else if (straight(hand))
                return (STRAIGHT);
            else
                return (HIGH_CARD);
        }
        /* If there are four unique card values, it can only be one pair */
        else if (unique_vals == 4) {
            return (ONE_PAIR);
        }
        /* If there are three unique card values, then either it's three of a kind or two pair */
        else if (unique_vals == 3) {
            /* pip_list [0].count has the count of the most frequently occuring pip */
            /* If most frequent unique value occurs three times, then it's three of a kind */
            if (how_often == 3)
                return (THREE_KIND);
            else
                return (TWO_PAIR);
        }
        /* If there are two unique card values, it is four of a kind or a full
           house (three and two of a kind) */
        else if (unique_vals == 2) {
            /* If most frequent unique value occurs four times, then it's four of a kind */
            if (how_often == 4)
                return (FOUR_KIND);
            else
                return (FULL_HOUSE);
        }
        /* Without wild cards you cannot have five of a kind */
        else
           printf ("Bad value for number of unique values: %d\n", unique_vals);
    }

int two_pair_tie_breaker(pip_list1, pip_list2)
    t_card_count pip_list1 [5], pip_list2 [5];
    /* Determine the winner when both hands have two pair. It's complicated enough
       to need its own function.
       :param pip_list1: array of struct with pip values and frequency counts for hand1
       :param pip_list2: array of struct with pip values and frequency counts for hand2
       :return: number of winning hand
     */
     {
        /* Check to see who has the highest two pair */
        if (pip_list1 [0].pip > pip_list2 [0].pip) {
            return 1;
        } else if (pip_list2 [0].pip > pip_list1 [0].pip) {
            return 2;
        /* Top pair is equal. Check second pair */
        } else if (pip_list1 [1].pip > pip_list2 [1].pip) {
            return 1;
        } else if (pip_list2 [1].pip > pip_list1 [1].pip) {
            return 2;
        /* Both pair are equal. Check fifth card (third in pip_list array) */
        } else if (pip_list1 [2].pip > pip_list2 [2].pip) {
            return 1;
        } else if (pip_list2 [2].pip > pip_list1 [2].pip) {
            return 2;
        } else {
            printf ("All cards match in two pairs. Illegal input\n");
            return -1;
        }
    }
        
int one_pair_tie_breaker(hand1, hand2, pip_list1, pip_list2)
    t_card hand1 [5], hand2 [5];
    t_card_count pip_list1 [5], pip_list2 [5];
    {
    /* Determine the winner when both hands have one pair. It's complicated enough
       to need its own function.
       :param hand1: an array of five elments of type card
       :param hand2: an array of five elments of type card
       :param pip_list1: array of struct with pip values and frequency counts for hand1
       :param pip_list2: array of struct with pip values and frequency counts for hand2
       :return: number of winning hand
     */
        
        int i;
        /* Check to see who has the higher pair */
        if (pip_list1 [0].pip > pip_list2 [0].pip) {
            return 1;
        } else if (pip_list2 [0].pip > pip_list1 [0].pip) {
            return 2;
        /* Pair has same value, check highest of each hand one by one until mismatch */
        } else {
            for (i = 0; i < 5; i++) {
                if (hand1[i].pip > hand2[i].pip) {
                    return 1;
                } else if (hand2[i].pip > hand1[i].pip) {
                    return 2;
                }
            }
        }
        /* All cards equal */
        printf ("Error one_pair_tie_breaker: All cards equal. Illegal input\n");
        return -1;
    }
        
int high_card_tie_breaker(hand1, hand2)
    t_card hand1 [5], hand2 [5];
    {
    /* Determine the winner when just high card. Has enought steps to need its own funtion.
       :param hand1: an array of five elments of type card
       :param hand2: an array of five elments of type card
       :return: number of winning hand
     */
        int i;
        
        for (i = 0; i < 5; i++) {
            if (hand1[i].pip > hand2[i].pip)
                return 1;
            else if (hand2[i].pip > hand1[i].pip)
                return 2;
        }
        /* All cards equal */
        printf ("Error high_card_tie_breaker: All cards equal. Illegal input\n");
        return -1;
    }
        
int tie_breaker (type_of_hand, hand1, hand2, pip_list1, pip_list2)
    int type_of_hand;
    t_card hand1 [5], hand2 [5];
    t_card_count pip_list1 [5], pip_list2 [5];
    /* Determine the winner when both hands have one pair. It's complicated enough
       to need its own function.
       :param type_of_hand: encoding of hand type
       :param hand1: an array of five elments of type card
       :param hand2: an array of five elments of type card
       :param pip_list1: array of struct with pip values and frequency counts for hand1
       :param pip_list2: array of struct with pip values and frequency counts for hand2
       :return: number of winning hand
     */
    {
        /* If the hand is a flush, straight, or straight flush */
        if ((type_of_hand == STRAIGHT_FLUSH) || (type_of_hand == FLUSH) ||
            (type_of_hand == STRAIGHT)) {
        /* Can determine by comparing the highest card in each hand */
            if (hand1[0].pip > hand2[0].pip) {
                printf ("Hand1 wins\n");
                return 1;
            } else if (hand2[0].pip > hand1[0].pip) {
                return 2;
            } else {
                printf ("Tie in hands - bad data\n");
                return -1;
            }
        }
        /* Hand is four of a kind */
        else if (type_of_hand == FOUR_KIND) {
            /* Card values in the four of a kind are unique. Just compare them */
            if (pip_list1 [0].pip > pip_list2 [0].pip) {
                return 1;
            } else {
                return 2;
            }
        }
        /* Hand is full house or three of kind */
        else if ((type_of_hand == THREE_KIND) || (type_of_hand == FULL_HOUSE)) {
        /* Can determine winner by comparing value of three of a kind. Must be different. */
            if (pip_list1 [0].pip > pip_list2 [0].pip) {
                return 1;
            } else {
                return 2;
            }
        }
        /* Two pair */
        else if (type_of_hand == TWO_PAIR) {
            return two_pair_tie_breaker (pip_list1, pip_list2);
        /* One pair */
        } else if (type_of_hand == ONE_PAIR) {
            return one_pair_tie_breaker (hand1, hand2, pip_list1, pip_list2);
        /* High card */
        } else if (type_of_hand == HIGH_CARD) {
            return high_card_tie_breaker (hand1, hand2);
        } else {
            printf ("Illegal hand type (over nine or less than one\n");
            return -1;
        }
    }
        
int main()
    {
    /* Main loop for Project Euler problem #54
        
       Sets-up main data structures
       Reads each line in the file, and converts into data structure format
       Calculates how often each pip value appears in each hand
       Calculates type and compares to determine winner, or if tie breaker is needed
     */

        int hand1_wins, hand1_type, hand2_type;
        int unique_vals, i;
        t_card hand1 [5], hand2 [5];
        t_card_count pip_list1 [5], pip_list2 [5];
        FILE *in_file;
        char *file_name = "poker.txt";
        char in_buffer [40];
        char *chr_ptr, *token;

        hand1_wins = 0;
        /* Each line is one pair of poker hands */
        in_file = fopen (file_name, "r");
        if (in_file == NULL) {
           printf ("Can't open file %s\n", file_name);
           return;
        }
        while (fgets (in_buffer, sizeof(in_buffer), in_file)) {
            chr_ptr = in_buffer;
            i = 0;
            while (token = strsep (&chr_ptr, " ")) {
                if (i < 5) { /* processing hand 1's cards */
                    hand1 [i].pip  = card_val_convert  (&token [0]);
                    hand1 [i].suit = card_suit_convert  (token [1]);
                } else { /* processing hand 2's cards */
                    hand2 [i-5].pip  = card_val_convert  (&token [0]);
                    hand2 [i-5].suit = card_suit_convert  (token [1]);
                }
                i++;
            }
            qsort (&hand1, 5, sizeof(t_card), hand_pip_compare);
            qsort (&hand2, 5, sizeof(t_card), hand_pip_compare);
            /* Figure out how often each pip value occurs */
            unique_vals = count_pips (hand1, pip_list1);
            hand1_type = hand_type (hand1, unique_vals, pip_list1 [0].count);
            unique_vals = count_pips (hand2, pip_list2);
            hand2_type = hand_type (hand2, unique_vals, pip_list2 [0].count);
            printf ("Hand 1 type: %d, Hand 2 type: %d\n", hand1_type, hand2_type);
            if (hand1_type > hand2_type) {
                printf ("Hand 1 wins\n");
                hand1_wins += 1;
            } else if (hand1_type == hand2_type) {
                if (tie_breaker (hand1_type, hand1, hand2, pip_list1, pip_list2) == 1) {
                    printf ("Hand 1 wins\n");
                    hand1_wins += 1;
                }
            }
            /* Put rest of processing in here.
            Need to do sort. Probably create the pip_list */
        }
        printf ("\nHand 1 wins %d times\n", hand1_wins);
        /*
        
        # Determine the winning hand
        # If different types of hands, the higher encoding wins
        else if hand1_type == hand2_type:
            # Hand types are the same. Call the tie breaking routine
            if tie_breaker(hand1_type, hand1, hand2, counts_card1, counts_card2) == 1:
                print("Hand 1 wins")
                hand1_wins += 1
    */
    }
